### PR TITLE
Allow for custom data-readers to deserialize cookies.

### PR DESCRIPTION
**Pull request by patrickthebold**
https://github.com/ring-clojure/ring/issues/302
This should give the user more control over how data is serialized in the session. However it will break some things since the edn reader defaults to [default-data-readers](https://github.com/clojure/clojure/blob/clojure-1.9.0-alpha14/src/clj/clojure/core.clj#L7625). So if someone was using one of those two defaults they would have to add them in to 'data_readers.clj' on the class path.

### CODE REVIEW

**Comment by weavejester** [comment]
I'd rather this be included as an option when the store is created, so people can supply either `*data-readers*` or their own readers map to the store.

**Comment by patrickthebold** [comment]
That's better, I'll make the update soonish.

File: ring-core/src/ring/middleware/session/cookie.clj [hunk 1]
```diff
@@ -66,35 +66,35 @@
     (random/bytes 16)))
 
 (defn- deserialize
-  [x]
-  (edn/read-string {:readers *data-readers*} x))
+  [options x]
```
Comment (weavejester) [inline ring-core/src/ring/middleware/session/cookie.clj]: Can you move `options` to the last argument for all these functions?

File: ring-core/src/ring/middleware/session/cookie.clj [hunk 2]
```diff
@@ -108,9 +108,15 @@
   :key - The secret key to encrypt the session cookie. Must be exactly 16 bytes
          If no key is provided then a random key will be generated. Note that in
          that case a server restart will invalidate all existing session
-         cookies."
+         cookies.
+
+  :readers - Passed to clojure.edn/read to read the serialized edn from the cookie.
```
Comment (weavejester) [inline ring-core/src/ring/middleware/session/cookie.clj]: Maybe something more like:

```
  :readers - A map of data readers used to read the serialized edn from the
             cookie. For writing, ensure that each data type has a key in the
             clojure.core/print-method or clojure.core/print-dup multimethods.
```

**Comment by patrickthebold** [comment]
Updates made as requested. Let me know if you need anything else to get this merged.

File: ring-core/src/ring/middleware/session/cookie.clj [hunk 3]
```diff
@@ -65,32 +65,36 @@
       secret-key)
     (random/bytes 16)))
 
-(defn- ^String serialize [x]
-  {:post [(= x (edn/read-string %))]}
+(defn- deserialize
+  [x options]
```
Comment (weavejester) [inline ring-core/src/ring/middleware/session/cookie.clj]: Can you put the arguments on the same line as the function name?

File: ring-core/test/ring/middleware/session/test/cookie.clj [hunk 4]
```diff
@@ -56,3 +56,16 @@
   (is (thrown-with-msg?
        AssertionError #"the secret key must be exactly 16 bytes"
        (cookie-store {:key (.getBytes "012345678901234567890")}))))
+
+; setup for serializing/deserializing Instant.
+(defmethod print-method java.time.Instant
+  [dt out]
+  (.write out (str "#foo/instant \"" (.toString dt) "\"")))
```
Comment (weavejester) [inline ring-core/test/ring/middleware/session/test/cookie.clj]: At the end of each `defmethod`, `defn` or `deftest` there should be a blank line.

File: ring-core/test/ring/middleware/session/test/cookie.clj [hunk 5]
```diff
@@ -56,3 +56,16 @@
...
+(defn parseInstant [x] (java.time.Instant/parse x))
+(deftest cookie-session-custom-type
+  (let [store    (cookie-store {:readers {'foo/instant #'ring.middleware.session.test.cookie/parseInstant}})
```
Comment (weavejester) [inline ring-core/test/ring/middleware/session/test/cookie.clj]: `#'ring.middleware.session.test.cookie/parseInstant` can be written more concisely as `#'parseInstant`

File: ring-core/test/ring/middleware/session/test/cookie.clj [hunk 6]
```diff
@@ -56,3 +56,16 @@
   (is (thrown-with-msg?
        AssertionError #"the secret key must be exactly 16 bytes"
        (cookie-store {:key (.getBytes "012345678901234567890")}))))
+
+; setup for serializing/deserializing Instant.
+(defmethod print-method java.time.Instant
+  [dt out]
+  (.write out (str "#foo/instant \"" (.toString dt) "\"")))
+(defn parseInstant [x] (java.time.Instant/parse x))
```
Comment (weavejester) [inline ring-core/test/ring/middleware/session/test/cookie.clj]: `parseInstant` should be `parse-instant`.

File: ring-core/test/ring/middleware/session/test/cookie.clj [hunk 7]
```diff
@@ -56,3 +56,16 @@
   (is (thrown-with-msg?
        AssertionError #"the secret key must be exactly 16 bytes"
        (cookie-store {:key (.getBytes "012345678901234567890")}))))
+
+; setup for serializing/deserializing Instant.
+(defmethod print-method java.time.Instant
+  [dt out]
```
Comment (weavejester) [inline ring-core/test/ring/middleware/session/test/cookie.clj]: Arguments should be on same line as `defmethod` if there's space.

**Comment by weavejester** [comment]
Thanks for the update. I've added some formatting corrections in a review. Could you also squash your commits into something like:

```clojure
Add :readers option to cookie-store
```

**Comment by deepxg** [comment]
Any chance this could make its way into a release some time soon?

**Pull request merged**